src/webgpu/shader/execution/builtin/builtin.ts

```diff
@@ -0,0 +1,258 @@
+  length: number
+): TypedArrayBufferView {
+  let inputData: TypedArrayBufferView;
+  switch (numberType) {
+    case OperandType.Float: {
+      inputData = new Float32Array(4 * cases.length);
+      break;
+    }
+    case OperandType.Hex:
+    case OperandType.Uint: {
+      inputData = new Uint32Array(4 * cases.length);
+      break;
+    }
+    case OperandType.Int: {
+      inputData = new Int32Array(4 * cases.length);
+      break;
+    }
+  }
+  for (let i = 0; i < cases.length; i++) {
+    for (let j = 0; j < length; j++) {
+      inputData[i * 4 + j] = cases[i].input;
+    }
+  }
+  return inputData;
+}
+
+export function submitComputeShader(
+  source: string,
+  t: GPUTest,
+  inputData: TypedArrayBufferView
+): GPUBuffer {
+  const inputBuffer = t.makeBufferWithContents(
+    inputData,
+    GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST | GPUBufferUsage.STORAGE
+  );
+
+  const outputBuffer = t.makeBufferWithContents(
+    new Float32Array(inputData.length),
+    GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST | GPUBufferUsage.STORAGE
+  );
+
+  const module = t.device.createShaderModule({ code: source });
+  const pipeline = t.device.createComputePipeline({
+    compute: { module, entryPoint: 'main' },
+  });
+
+  const group = t.device.createBindGroup({
+    layout: pipeline.getBindGroupLayout(0),
+    entries: [
+      { binding: 0, resource: { buffer: inputBuffer } },
+      { binding: 1, resource: { buffer: outputBuffer } },
+    ],
+  });
+
+  const encoder = t.device.createCommandEncoder();
+  const pass = encoder.beginComputePass();
+  pass.setPipeline(pipeline);
+  pass.setBindGroup(0, group);
+  pass.dispatch(1);
+  pass.endPass();
+
+  t.queue.submit([encoder.finish()]);
+  return outputBuffer;
+}
+
+export function createBuiltinCall(
+  baseType: ScalarType,
+  arrayLength: number,
+  builtin: string,
+  outputDataAsF32: Float32Array,
+  inputData: TypedArrayBufferView,
+  cases: Cases,
+  numberType: OperandType
+): string[] {
+  let inputDataAsF32: Float32Array;
+  let inputDataAsU32: Uint32Array;
+  let inputDataAsI32: Int32Array;
+  if (numberType === OperandType.Float) {
+    inputDataAsF32 = inputData as Float32Array;
+    inputDataAsU32 = new Uint32Array(inputData.buffer);
+    inputDataAsI32 = new Int32Array(inputData.buffer);
+  } else {
+    inputDataAsF32 = new Float32Array(inputData.buffer);
+    inputDataAsU32 = inputData as Uint32Array;
+    inputDataAsI32 = new Int32Array(inputData.buffer);
+  }
+
+  const outputDataAsU32 = new Uint32Array(outputDataAsF32.buffer);
+  const outputDataAsI32 = new Int32Array(outputDataAsF32.buffer);
+
+  const errs: string[] = [];
+  for (let i = 0; i < cases.length; i++) {
+    const input: string[] = [];
+    const output: string[] = [];
+    const expected: string[] = [];
+    let matched = true;
+    for (let j = 0; j < arrayLength; j++) {
+      const idx = i * 4 + j;
+      const caseExpected: string[] = [];
+      cases[i].expected.forEach((e: number) => {
```

Comment:
nit: prefer a for-each loop: `for (const e of cases[i]) {`

Reply:
done